### DISCUSSION TITLE
Confusion with custon operator definition

### DISCUSSION THREAD

**Discussion by anubhavkamal** [MilesCranmer/PySR#522] [Q&A]
def p(i):
    if i < 20.0:
        return sympy.log(27.1285645162762 - 1.42277605519119*i)
    else:
        return float("nan")
model = pysr.PySRRegressor(
        turbo=True,
        maxsize=30,
        batching=True,
        warm_start=True,
        model_selection='best',
        niterations=40000,
        precision = 64,
        binary_operators=['+','-','*','/'],
        #unary_operators=['exp', 'log'],
        unary_operators = ['exp', "fn(x) = log(27.1285645162762 - 1.42277605519119*x)"],
        extra_sympy_mappings = {"fn": lambda x: p(x)},
        nested_constraints = {"exp": {"exp" : 0}},
        #loss = 'LogitDistLoss()',
        )
        
        
        
       Hello!
       I have defined a custom function of the form shown above but PySR is throwing following error - 
       
       RuntimeError: <PyCall.jlwrap (in a Julia function called from Python)
JULIA: The operator `fn` is not well-defined over the real line, as it threw the error `DomainError` when evaluating the input 20.408163265306122. You can work around this by returning NaN for invalid inputs. For example, `safe_log(x::T) where {T} = x > 0 ? log(x) : T(NaN)`.

The inputs in the dataset is less than 18. Thus, there is no case where the value of log(x) blows to infinity. I am not sure about this 'DomainError'.

**Comment by MilesCranmer** [comment]
> The inputs in the dataset is less than 18. Thus, there is no case where the value of log(x) blows to infinity. I am not sure about this 'DomainError'.

Keep in mind it’s going to try many different things as input to your operator; not just the dataset feature itself. Because you have `/` it could also scale your features. So you always want to catch values before they cause a domain error and return NaN instead.

cheers,
Miles

**Reply by anubhavkamal** [reply]
Hello!
Thanks for the swift reply.

While defining other "questionable" functions such as 'log(x)', '1/log(x)', '1/x'. I didn't face this domainError. 
What shall I change here, if I want a particular structure in my function?
Thanks!

**Reply by MilesCranmer** [reply]
Some functions get automatically mapped to "safe" versions. Like if you write `log` as one of the operators, it will get automatically converted to `x > 0 ? log(x) : NaN`. But for custom functions this conversion doesn't happen.

**Reply by anubhavkamal** [reply]
Thanks @MilesCranmer !
Turns out to be an easier fix. Just used 'absolute' value function before passing it to the log function.
Your answers helped!
